### PR TITLE
msg: we should set the socket options before connect or listen in order to have it take effect.

### PR DESCRIPTION
**Pull request by yangruifeng**
Signed-off-by: Ruifeng Yang <149233652@qq.com>

### CODE REVIEW

**Comment by yangruifeng** [comment]
@tchaikov sorry, i have not digged the implement of socket, so i am not sure. in my experience,  TCP_NODELAY and SO_RCVBUF take effect on the process of connect(), but do not affect the result of connect(). test is ok.

**Comment by yuyuyu101** [comment]
@tchaikov This commit make nonblock connect also set socket options. Otherwise like now it will fast return and skip set_socket_options

File: src/msg/async/net_handler.cc [hunk 1]
```diff
@@ -126,8 +129,6 @@ int NetHandler::generic_connect(const entity_addr_t& addr, bool nonblock)
     return -errno;
   }
 
-  set_socket_options(s);
-
```
Comment (yuyuyu101) [inline src/msg/async/net_handler.cc]: Hmm, but we already have set_socket_options at AsyncConnection.cc. Maybe we can just remote this line.

Comment (yangruifeng) [inline src/msg/async/net_handler.cc]: @yuyuyu101  you are right, i will change the title, and resubmit the pull request.

Comment (yangruifeng) [inline src/msg/async/net_handler.cc]: @yuyuyu101 @tchaikov 
after testing(ss command line), I found that SO_RCVBUF has no effect, and we should set it before connect or listen.

man 7 tcp, get the information:
The maximum sizes for socket buffers declared via the SO_SNDBUF and SO_RCVBUF mechanisms are limited by the values in the /proc/sys/net/core/rmem_max and /proc/sys/net/core/wmem_max  files.
Note  that  TCP  actually  allocates  twice  the  size  of  the buffer requested in the setsockopt(2) call, and so a succeeding getsockopt(2) call will not return the same size of buffer as requested in the setsockopt(2) call.  TCP uses the extra space for administrative purposes and internal kernel structures, and the /proc file values reflect the larger sizes compared to the actual TCP windows.  On individual connections, **the socket buffer size must be set prior to the listen(2) or connect(2) calls in order to have it take effect.**  See socket(7) for more information.

Comment (tchaikov) [inline src/msg/async/net_handler.cc]: @yangruifeng, thanks for testing and offering man page. so you are right! we need to set `RCVBUF` **before** listen(2).

**Comment by yangruifeng** [comment]
@yuyuyu101 @tchaikov 
please check again, thanks.

**Comment by yuyuyu101** [comment]
Reviewed-by:

**Comment by tchaikov** [comment]
lgtm with qa run.

**Comment by tchaikov** [comment]
tested at 
- http://pulpito.ceph.com/kchai-2015-09-08_02:39:25-rados-wip-kefu-testing---basic-multi/
- http://pulpito.ceph.com/kchai-2015-09-08_10:19:56-rados-wip-kefu-testing---basic-multi/
- http://pulpito.ceph.com/kchai-2015-09-08_18:05:34-rados-wip-kefu-testing---basic-multi/

**Pull request merged**